polaris/polaris/management/commands/poll_pending_deposits.py

```diff
@@ -138,17 +125,96 @@ def execute_deposits(cls):
         for transaction in ready_transactions:
             if module.TERMINATE:
                 break
+            elif transaction.amount_in is None:
+                raise CommandError(
+                    "poll_pending_deposits() did not assign a value to the "
+                    "amount_in field of a Transaction object returned"
+                )
+            elif transaction.amount_fee is None:
+                if registered_fee_func == calculate_fee:
+                    transaction.amount_fee = calculate_fee(
+                        {
+                            "amount": transaction.amount_in,
+                            "operation": settings.OPERATION_DEPOSIT,
+                            "asset_code": transaction.asset.code,
+                        }
+                    )
+                else:
+                    transaction.amount_fee = Decimal(0)
             try:
-                success = execute_deposit(transaction)
-            except ValueError as e:
-                logger.error(str(e))
+                _, created = get_or_create_transaction_destination_account(transaction)
+            except RuntimeError as e:
+                transaction.status = Transaction.STATUS.error
+                transaction.status_message = str(e)
+                transaction.save()
+                logger.error(transaction.status_message)
                 continue
-            if success:
-                # Get updated status
-                transaction.refresh_from_db()
+            if created:
+                # Transaction.status == pending_trust, wait for client
+                # to add trustline for asset to send
+                continue
+            master_signer = None
+            if transaction.asset.distribution_account_master_signer:
+                master_signer = json.loads(
+                    transaction.asset.distribution_account_master_signer
+                )
+            thresholds = json.loads(transaction.asset.distribution_account_thresholds)
+            if not (
+                master_signer and master_signer["weight"] >= thresholds["med_threshold"]
+            ):
+                # master account is not sufficient for
+                transaction.pending_signatures = True
+                transaction.status = Transaction.STATUS.pending_anchor
+                transaction.save()
+                if transaction.channel_account:
```

Comment:
@JakeUrban Maybe the channel account should be obtained only from calling `rdi.channel_keypair_for_multisig_transaction`. It can be confusing to be able to set `transaction.channel_account` and also have an integration function to provide it.
Polaris should be the one setting `transaction.channel_seed` with the value obtained from the integration function.
This also makes it easier for the Anchor to handle channel accounts, since the entire logic can reside inside the integration function, no need to worry about storing it in the transaction beforehand.

Reply:
Or maybe this is just to make it easier for writing tests, since you can set `channel_seed` directly. Nevermind the comment.

Reply:
I actually agree. `rdi.channel_keypair_for_multisig_transaction` will not be passed a `transaction` object now. Instead it should return a new/fresh channel account every time it's called. Maybe it should just be called `rdi.create_channel_account()`?

Reply:
Not that there are only two times a channel account would be used per-transaction:
- When the deposit transaction is submitted (obviously)
- But also before that, if the destination account doesn't exist, a channel account will be used as the source account of the TX while the asset's distribution account is the source of the payment operation.

Since this 2nd (1st if going in-order of the flow) channel account isn't actually for the deposit transaction, `rdi.after_deposit()` is never called for that channel account. So I think we need a clean-up function to merge outstanding channel accounts.

`rdi.merge_channel_account()`? Called after Polaris submits the TX for which the channel account was created for.

Reply:
Then `rdi.after_deposit()` wouldn't be the place you merge channel accounts anymore.

Reply:
Or actually, maybe we continue with one channel account per transaction, not two. If we choose to go that route, then `transaction.channel_account` makes sense to have because it will always be used for transactions related to the actual deposit transaction the object represents.

We'd still have `create_channel_account()` and `merge_channel_account()`, but they would be called when the first TX needs to be submitted on behalf of the deposit and when the deposit has completed or failed, respectively. 

If the deposit failed, it should be the anchor if they want to merge the channel account back or wait until the deposit is retried after resolving an issue. When a previously failed deposit is retried after manual intervention, `merge_channel_account()` would be called again.

Reply:
Would it make more sense for `create_channel_account()` not to return anything and instead be passed the `transaction` object, so the anchor could assign `transaction.channel_seed`? Polaris would still validate the existence of the account.

Reply:
`merge_channel_account` lines up exactly with `after_deposit()` so we're not going to do `merge_channel_account()` and instead there will be added emphasis on `after_deposit()`.

Reply:
@JakeUrban `rdi.create_channel_account()` would require on-demand multi-sig, which is what we're trying to avoid.
Since the need for channel accounts on this multi-sig feature is due to the idea of allowing the Anchor do polling instead of on-demand, I think we should drop the polling support and do on-demand multi-sig only, in which the Anchor **has** to sign the envelope when Polaris tries to submit it. Otherwise it's a circular logic and sequence numbers will always be an issue.

Reply:
I agree it would be simpler to require a push method but I think thats pretty restrictive.

I'm not sure what you mean by "on-demand multisig". As described in some comments I have locally (changes I haven't pushed):
```
    If the transacted asset's distribution account does require non-master signatures, the anchor
    should save a keypair of a pre-existing Stellar account to use as the channel account via
    DepositIntegration.create_channel_account().

    This channel account must only be used as the source account for transactions related to the
    ``Transaction`` object passed. It also must not be used to submit transactions by any service
    other than Polaris. If it is, the outstanding transactions will be invalidated due to bad
    sequence numbers. Finally, the channel accounts must have a master signer with a weight greater
    than or equal to the medium threshold for the account.
```

Reply:
This last line: `Finally, the channel accounts must have a master signer with a weight greater than or equal to the medium threshold for the account` is important. 

It means the asset's distribution account requires multisig, but the channel accounts intentionally don't. Channel accounts should have master account that can sign for its medium-threshold transactions. This is totally fine because the actual deposit transaction still requires the distribution account's signatures. Meaning, its not any less secure.

The transaction that creates the destination account, if it doesn't exist, is the one that doesn't require the distribution account's signatures, because the channel account is paying for the created account's XLM reserve. This is also not a security concern because the channel account is created on-the-fly per-transaction. It should not have a significant balance of _any_ asset and should be destroyed as soon as the deposit has been facilitated.

Does this address your issues with channel accounts?

Reply:
@JakeUrban 

> I agree it would be simpler to require a push method but I think thats pretty restrictive.

Me too.

> I'm not sure what you mean by "on-demand multisig"

On-demand means that Polaris needs the return value right now. For example, if the multi-sig for deposit was on-demand, the Anchor would have to add all signatures in a single function call ("push" method as you described previously) so Polaris could submit the envelope to horizon instantly.

> the anchor should save a keypair of a pre-existing Stellar account to use as the channel account

If the keypair is per-transaction, then I imagine the Anchor would have to create it somewhere like `process_sep6_deposit()`, and on this case it would have to add all signatures _on-demand_ (as described above) and store the keypair somewhere attached to this transaction, likely in a related model. That's why I said it would require on-demand multi-sig, and if the Anchor has, at some point, to do on-demand multi-sig, then I don't see why it wouldn't do on-demand for all cases, inclusing sending the payment to the user.

However, I do see a scenario where there's no need for on-demand multi-sig. The Anchor can simply not allow Polaris to process the transaction until it creates the channel account. This means the Anchor doesn't return the transaction in `poll_pending_deposits()` until there's a channel account ready for that transaction.

Reply:
> If the keypair is per-transaction, then I imagine the Anchor would have to create it somewhere like process_sep6_deposit(), and on this case it would have to add all signatures on-demand (as described above) and store the keypair somewhere attached to this transaction, likely in a related model. 

All anchors, regardless of if they're implementing SEP-6 or SEP-24, will implement `create_channel_account(transaction)`, which should create the channel account on Stellar. It should be simple, non-multisig account. 

Polaris will call this function when one of two things happen, assuming the user sent the off-chain deposit to the anchor and the anchor has detected/matched that deposit with the internal record:
 - If the client's Stellar account doesn't exist, `create_channel_account()` is called to create it. The channel account is the one that pays the XLM reserve for the account created. This channel account must be a simple master-signer account, AKA no multisig.
- If the clients Stellar account does exist, then `create_channel account()` will be called for `transaction` so it can be used to make the actual deposit transaction to the client's Stellar account

So a channel account will be used for 1-2 transactions, depending on if the Stellar account exists. If the asset's distribution account is not configured for multisg, channel accounts will not be used.

Since the channel account is per-transaction, the channel account will be stored on `Transaction.channel_seed` and `Transaction.channel_account`.

Reply:
Let me write up the documentation and implement it in the anchor server so you can see how it would work for SEP-24 and SEP-6. I think we have what we need now.

Reply:
I think the bottom line in response to your concern is that all channel accounts _must_ be created on-demand and merged back to whatever account is creating them when the TX has been successfully submitted on the network.

Reply:
@JakeUrban 

> All anchors, regardless of if they're implementing SEP-6 or SEP-24, will implement create_channel_account(transaction), which should create the channel account on Stellar. It should be simple, non-multisig account.

Yes, but to create the channel account, multiple signers are required. At least that's what I understood from this statement:

> I'm working on implementing the channel account functionality and I'm realizing that if an asset's distribution account requires multiple signatures for payment transactions, the anchor must collect signatures for the transaction that creates the channel account.

So, in `create_channel_account()`, the Anchor has to instantly collect all signatures otherwise it won't be able to create the channel account. Is that correct?

Reply:
I might be a bit lost on the restrictions we have with multi-sig. This is my current understanding:
1) to avoid issues with sequence numbers, we need channel accounts
2) the asset distribution account must create the channel account for each transaction, because the channel account must be derived from the distribution account
3) if the distribution account required multi-sig for its transaction, then it needs to collect signatures in order to **create** the channel account
4) once the channel account, is created, it will be used to send payments to the user and will be the source account of all the payments
  * when the channel account creates the envelope to send the payment to the user, does is require multi-sig on that envelope?

I'm not sure exactly where multiple signatures are required in all this process, that's why my ideas might seem a bit over complicated.

Reply:
Close, but yes you are overcomplicating it. I also didn't help by making comments as I think through the process myself.

> Yes, but to create the channel account, multiple signers are required.
> 2. the asset distribution account must create the channel account for each transaction, because the channel account must be derived from the distribution account

These are both false, the distribution account does not need to be derived. It can be an independent account that is not created from the asset's distribution account. This gets around the need for multiple signers in order to create the channel account. 

Originally, I thought it would make sense to use derived accounts and use the distribution account to create the destination account, but when I realized the issue of still needing signatures, I decided to remove that constraint.

Reply:
> when the channel account creates the envelope to send the payment to the user, does is require multi-sig on that envelope?

It will need multisig if the distribution account is used to send the payment.

Reply:
@JakeUrban Thanks for the clarification. Please correct me if I'm wrong:
1. Polaris needs a source account to send payments
2. Anchor wants to require multiple signatures to send payments
3. To allow multiple signers on a transaction, Polaris needs to store the envelope and let the Anchor sign it whenever it wants
4. (3) Introduces the issue with sequence numbers
5. To work around the issue with sequence numbers, Polaris allows the Anchor to use an individual stellar account for each transaction, which is called channel account
6. The Anchor, for each transaction, would create and fund a channel account (which is temporary and will be merged back after the payment is sent?)
7. Polaris, when creating the deposit envelope, checks if the asset distribution account requires multiple signers. If yes, it sets `pending_signatures = True` and relies on the Anchor to provide a channel account for the transaction. The Anchor must provide a channel account instantly (see 7.3 below) when `rdi.create_channel_account()` is called.
- 7.1. Is it possible for the Anchor to create a channel account that doesn't require multiple signers to send payments?
- 7.2. If (7.1) is true, then I imagine that, the envelope that creates the channel account needs to be signed by the signers required initially on (2), correct?
- 7.3. Since the channel account must be instantly created and returned in `rdi.create_channel_account()`, and if (7.2) is true, then inside `rdi.create_channel_account()` the Anchor must implement some mechanism to do multi-signatures by requesting them using the push model, correct?
  * Push models that require the anchor to send the envelope to signing parties, the signing parties would receive, sign, and return the transaction envelope to the anchor service
- 7.4. If (7.3) is true, then Polaris would be requiring instant ("on-demand") multi-sig on `rdi.create_channel_account()` but then allowing "not instant" multi-sig on `transaction.envelope`. That's the scenario I was referring to in here:
  * If the keypair is per-transaction, then I imagine the Anchor would have to create it somewhere like process_sep6_deposit(), and on this case it would have to add all signatures on-demand (as described above) and store the keypair somewhere attached to this transaction, likely in a related model. That's why I said it would require on-demand multi-sig, and if the Anchor has, at some point, to do on-demand multi-sig, then I don't see why it wouldn't do on-demand for all cases, including sending the payment to the user.

Reply:
1 through 7 are completely accurate (yes to 6).

> 7.1. Is it possible for the Anchor to create a channel account that doesn't require multiple signers to send payments?

It is a _requirement_ that channel accounts have a master signer whose weight is greater than it's medium threshold. That means that if the channel account wants to send XLM (or any asset) _from its own account balance_, it can do so by signing the transaction with the single account public key secret.

I know this is probably confusing because we're using channel accounts specifically for multisig transactions. So when do transactions using channel accounts need multiple signatures?

Every Stellar transaction has a source account. In this context, this is the channel account we're talking about, _which only has the master signer on the account_. 

**Optionally, many (all?) Stellar operations within a transaction can have their own source account that differs from the transaction-level source account.**

This is what we would do for deposit (payment) transactions in SEP-6 and SEP-24 when the asset's distribution account requires multiple signatures. 

Let's say asset MULT requires 2 signatures: one of which is the master signature and the other is an extra signer required to reach the medium threshold. This deposit transaction envelope must have the following structure before submitting:

TransactionEnvelope

- source: `channel account public key`
- operations:
    - payment:
        - source: `asset distribution account public key`
        - asset_code: `MULT`
        - asset_issuer: `asset issuer account`
- signatures:
    - `channel account master key signature`
    - `asset distribution account master key signature`
    - `asset distribution account extra signer key signature`

And remember, creating the channel account for a multisig transaction should _not_ be done by the distribution account, because the distribution account requires multiple signatures. Channel accounts must be created as a result of the `rdi.create_channel_account(tranaction)` function call, so use an account with a single signer to fund channel accounts.

I would set up a system where funds are sent from the distribution account to a _funding_ account periodically, say every hour. This funding account should have a single master signer so it can fund channel accounts for Polaris on-demand. It could also be the account funds are merged back to after the deposit completed (or funds could be merged directly back to the distribution account).

There are other ways you could setup accounts so that channel accounts can be created on-demand, but the idea is you need channel accounts to be created on demand from `create_channel_account()` and presumably this must be done with a non-multisig funding account, AKA not the distribution account.

Let me know if you have any questions about the above ^. I'm going to talk now about how channel accounts are _sometimes_ used twice per-deposit transaction.

**There's a special case: when the destination account doesn't exist**

If the destination account for the Stellar deposit doesn't exist, the anchor should create (sponsor) the account for them. This is typically done with the asset's distribution account funds. But when distribution accounts require multisig, the channel account for the transaction can be created early and used to fund the user's destination account.

So when the destination account for the deposit doesn't exist, a transaction like so is submitted:

TransactionEnvelope

- source: `channel account public key`
- operations:
    - create_account:
        - source: `channel account public key`
        - starting_balance: `minimum XLM starting balance` (2.01?)
- signatures:
    - `channel account master key signature`

Note that the transaction doesn't need multiple signatures. Thats because this transaction doesn't involve an account that requires multiple signatures.

So the channel account is used for this purpose and is kept around for when it needs to be used for the actual payment transaction to the destination account. Again, once payments are made, the channel account is merged back using `after_deposit()`. It should be merged back to the distribution account or the account that funded the channel account.

Hopefully I've addressed your questions and outlined how this can work. I'm curious to hear your thoughts. Obviously, if you want to implement a push model (immediate signing my additional signers), you can collect signatures quickly by calling an internal non-Polaris function and changing the `transaction.pending_signatures` back to `False` after Polaris makes `transaction.pending_signatures = True`.

Reply:
@JakeUrban I think I got the idea. The key information was about the transaction envelope source, which can differ from the source of each operation in the envelope.
So the funding account just needs to have a minimal XLM balance (doesn't even need other balances) for creating channel accounts on-demand.
The payment envelope signatures can be added via polling by the Anchor, as previously planned (when channel accounts weren't involved).
I really appreciate the patience and time to write up the explanation, thanks a lot! :)

Reply:
No worries! I want to make sure I get this use case right. One thing you mentioned just now:

> The payment envelope signatures can be added via polling by the Anchor, as previously planned (when channel accounts weren't involved).

I think you're referring to the poll model the anchor could implement so clients can detect and add signatures. But you don't necessarily have to do that. You can have a push model where you have a separate `notify_signers` process that checks for transactions in `pending_signature=True` and makes API calls or does whatever it needs to do to get the signatures immediately before returning.

Its up to you what the process is for collecting signatures, Polaris just creates the initial envelope and waits for the transaction to not be `pending_signatures=True` anymore.